File: MCL_by_python/Drive.py
import math
import InitializeMap
import Drive

global driveCommand, RobotPose

driveCommand = [
    # 右：90  左：-90
    #第一阶段
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    # [20, 0, 3],
    [20, -30, 1],
    [20, -30, 1],
    [20, -30, 1],
    #第二阶段
    # [20, 0, 1],
    # [20, 0, 1],
    # [20, 0, 1],
    [20, 0, 1],
    # [20, 0, 4],
    [20, 30, 1],
    [20, 30, 1],
    [20, 30, 1],
    # [20, 0, 3],
    #第三阶段
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    #第四阶段
    [20, 30, 1],
    [20, 30, 1],
    [20, 30, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    # [20, 0, 1],
    # [20, 0, 1],
    # [20, 0, 1],
    # [20, 0, 1],
    #第五阶段
    [20, -30, 1],
    [20, -30, 1],
    [20, -30, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    # [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    #第六阶段
    [20, 30, 1],
    [20, 30, 1],
    [20, 30, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    #第七 掉头
    [0, -30, 1],
    [0, -30, 1],
    [0, -30, 1],
    [0, -30, 1],
    [0, -30, 1],
    [0, -30, 1],
    #第八
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],
    [20, 0, 1],

    [20, 90, 1],
    [20, 0, 3],
    [20, 0, 4],
    [20, 90, 1],
    [20, 0, 4],
    [20, 0, 3],
    [20, 0, 3],
    [20, -90, 1],
    [20, 0, 3],
    [20, 0, 3],
    [20, 0, 3],
    [20, 0, 3],
    [20, 0, 3],
    [20, 0, 3],
    [20, 0, 3]
]

RobotPose = [[0] * 3 for i in range(len(driveCommand) + 1)]
RobotPose[0] = [50, 230, 0]


def MoveExact(i):
    global driveCommand, RobotPose

    if i == -1:

        return RobotPose[0][0], RobotPose[0][1], RobotPose[0][2]

    else:
        currentX = RobotPose[i][0]
        currentY = RobotPose[i][1]
        headRad = math.radians(RobotPose[i][2])
        if driveCommand[i][1] != 0:
            vOverOmegaRad = driveCommand[i][0] * 1.0 / math.radians(driveCommand[i][1])
            newX = currentX - vOverOmegaRad * math.sin(headRad) + vOverOmegaRad * math.sin(
                headRad + math.radians(driveCommand[i][1]) * driveCommand[i][2])
            newY = currentY + vOverOmegaRad * math.cos(headRad) - vOverOmegaRad * math.cos(
                headRad + math.radians(driveCommand[i][1]) * driveCommand[i][2])
            newThetaRad = headRad + math.radians(driveCommand[i][1]) * driveCommand[i][2]
        else:
            newThetaRad = headRad
            distanceTravel = driveCommand[i][0] * driveCommand[i][2]
            newX = currentX + distanceTravel * math.cos(headRad)
            newY = currentY + distanceTravel * math.sin(headRad)
        RobotPose[i + 1][0] = newX
        RobotPose[i + 1][1] = newY
        RobotPose[i + 1][2] = math.degrees(newThetaRad)
        print(newX, newY, newThetaRad)
        InitializeMap.DrawRobot(newX, newY, math.degrees(newThetaRad))

        return newX, newY, math.degrees(newThetaRad)

File: MCL_by_Cpp/MCL_C++/Drive.cpp
#include "stdafx.h"
#include <iostream>
#include <cmath>
#include "Drive.h"
#include "Points.h"
#include "Map.h"


Move DriveCommand[25] = {
  { 5,0,1 },
  { 5,0,1 },
  { 5,0,1 },
  { 5,10,1 },
  { 5,10,1 },
  { 5,10,1 },
  { 5,10,1 },
  { 5,10,1 },
  { 5,10,1 },
  { 5,10,1 },
  { 5,10,1 },
  { 5,10,1 },
  //{ 5,30,1 },
  //{ 5,30,1 },
  { 5,0,1 },
  { 5,-30,1 },
  { 5,-30,1 },
  { 5,-30,1 },
  { 5,0,1 },
  { 5,0,1 },
  { 5,0,1 },
  { 5,0,1 },
  { 5,0,1 },
  { 5,0,1 },
  { 5,0,1 },
  { 5,0,1 },
  { 5,0,1 },
};

int loop = -1;

int DCLen = sizeof(DriveCommand)/sizeof(Move);

Pose RobotPose[30] = { 0 };



/*************************************************
// Method: MoveExact
// Description: Move Robot
// Author: RSN
// Date: 2019/05/19
// Returns: Pose (Position after move)
// Parameter: step 
// History: 
*************************************************/
Pose MoveExact(int step)
{
  Pose newRP; // new Robot Position
  
  if (step == -1)
  {
    newRP.x = RobotPose[0].x;
    newRP.y = RobotPose[0].y;
	newRP.theta = RobotPose[0].theta;
    return newRP;
  }
  else
  {
    double currentX = RobotPose[step].x;
    double currentY = RobotPose[step].y;
	double newThetaRad = ang2rad(RobotPose[step].theta);
    
    if (DriveCommand[step].omega != 0)
    {
      double vOverOmegaRad = DriveCommand[step].v * 1.0 / ang2rad(DriveCommand[step].omega);
      newRP.x = currentX - vOverOmegaRad * sin(newThetaRad) 
                  + vOverOmegaRad * sin(newThetaRad + ang2rad(DriveCommand[step].omega) * DriveCommand[step].time);
      newRP.y = currentY + vOverOmegaRad * cos(newThetaRad) 
                  - vOverOmegaRad * cos(newThetaRad + ang2rad(DriveCommand[step].omega) * DriveCommand[step].time);
	  newThetaRad = newThetaRad + ang2rad(DriveCommand[step].omega) * DriveCommand[step].time;
    }
    else
    {
      double distanceTravel = DriveCommand[step].v * DriveCommand[step].time;
      newRP.x = currentX + distanceTravel * cos(newThetaRad);
      newRP.y = currentY + distanceTravel * sin(newThetaRad);
    }

    RobotPose[step+1].x = newRP.x;
    RobotPose[step+1].y = newRP.y;
    RobotPose[step+1].theta = rad2ang(newThetaRad);
    DrawRobot(newRP.x, newRP.y);
    
    newRP.theta=rad2ang(newThetaRad);
    return newRP;
   }
}
